src/main/java/org/prebid/server/auction/ExchangeService.java

```diff
@@ -463,21 +478,55 @@ private BidderSeatBid validateAndUpdateResponse(BidderSeatBid bidderSeatBid) {
     }
 
     /**
-     * Applies correction to {@link Bid#price}
-     * <p>
-     * Should be used after {@link BidderSeatBid} has been validated
-     * by {@link ExchangeService#validateAndUpdateResponse(BidderSeatBid)}
-     */
-    private static BidderSeatBid applyBidPriceAdjustment(BidderSeatBid bidderSeatBid,
-                                                         BigDecimal priceAdjustmentFactor) {
-        if (priceAdjustmentFactor != null) {
-            for (final BidderBid bidderBid : bidderSeatBid.getBids()) {
-                final Bid bid = bidderBid.getBid();
-                bid.setPrice(bid.getPrice().multiply(priceAdjustmentFactor));
-            }
+     * Performs changes on {@link Bid}s price depends on different between adServerCurrency and bidCurrency,
+     * and adjustment factor. Will drop bid if currency conversion is needed but not possible.
+     */
+    private BidderSeatBid applyBidPriceChanges(BidderSeatBid bidderSeatBid,
+                                               Map<String, Map<String, BigDecimal>> requestCurrencyRates,
+                                               String adServerCurrency, BigDecimal priceAdjustmentFactor) {
+        if (bidderSeatBid.getBids().isEmpty()) {
+            return bidderSeatBid;
         }
 
-        return bidderSeatBid;
+        final List<BidderError> errors = new ArrayList<>(bidderSeatBid.getErrors());
+
+        final List<BidderBid> updatedBidderBids = bidderSeatBid.getBids().stream()
```

Comment:
This stream pipeline is better to be converted to `foreach` block containing all conversions and error handling - it will be much easier to reason about error cases when that is done.